src/persist-client/src/internal/maintenance.rs

```diff
@@ -196,38 +207,47 @@ where
         V: Debug + Codec,
         D: Semigroup + Codec64 + Send + Sync,
     {
-        for future in self.perform_in_background(machine, gc, compactor) {
-            let _ = future.await;
-        }
+        self.perform_in_background(machine, gc, compactor).await
     }
 
     /// Initiates maintenance work in the background, either through spawned tasks
     /// or by sending messages to existing tasks. The returned futures may be
     /// awaited to know when the work is completed, but do not need to be polled
     /// to drive the work to completion.
-    fn perform_in_background<K, V, D>(
+    async fn perform_in_background<K, V, D>(
         self,
         machine: &Machine<K, V, T, D>,
         gc: &GarbageCollector<K, V, T, D>,
         compactor: Option<&Compactor<K, V, T, D>>,
-    ) -> Vec<BoxFuture<'static, ()>>
-    where
+    ) where
         K: Debug + Codec,
         V: Debug + Codec,
         D: Semigroup + Codec64 + Send + Sync,
     {
-        let mut futures = self.routine.perform_in_background(machine, gc);
+        let Self {
+            routine,
+            compaction,
+        } = self;
+        let mut more_maintenance = RoutineMaintenance::default();
+        for future in routine.perform_in_background(machine, gc) {
+            more_maintenance.merge(future.await);
+        }
 
         if let Some(compactor) = compactor {
-            for req in self.compaction {
+            for req in compaction {
                 if let Some(receiver) = compactor.compact_and_apply_background(req, machine) {
                     // it's safe to ignore errors on the receiver. in the
                     // case of shutdown, the sender may have been dropped
-                    futures.push(receiver.map(|_| ()).boxed());
+                    let _ = receiver.await;
```

Comment:
should compaction also return maintenance like `RoutineMaintenance` so it can all be merged together?

Reply:
Not sure! What do you think?

When I was drafting this PR up I mostly kicked off maintenance locally where it was possible, and only started threading it around when it couldn't be managed locally for whatever reason. But I don't think that was a very reasoned strategy; mostly it seemed consistent with how it was done elsewhere.

Reply:
fwiw, I don't think it gets us much unless we want to do something like make the contract of this method that all (potentially recursive) maintenance is finished by the time it returns (i.e. no `perform_in_background` calls). which seems like it could be a cool idea, but I wouldn't hold up merging over it